### PR TITLE
[WFLY-8136] allow JAX-RS annotations in CDI @Stereotype

### PR DESCRIPTION
**Pull request by xstefank**
https://issues.jboss.org/browse/WFLY-8136

### CODE REVIEW

**Comment by asoldano** [comment]
@ronsigal , please see here

File: jaxrs/WFLY-8136_EAP7-1228_JAX-RS_annotations_in_CDI_Stereotype.adoc [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+:toc:               left
+:icons:             font
+:idprefix:
+:idseparator:       -
+
+== Overview
+
+This feature provides support for JAX-RS annotations to be defined as a part of 
+CDI `@Stereotype` definitions. This will allow JAX-RS annotations like `@Produces` 
+or `@Consumes` to be reused by several JAX-RS resources or providers.
```
Comment (marekkopecky) [inline jaxrs/WFLY-8136_EAP7-1228_JAX-RS_annotations_in_CDI_Stereotype.adoc]: > like `@Produce` or `@Consume`

Can you please provide full list of supported annotations here?

File: jaxrs/WFLY-8136_EAP7-1228_JAX-RS_annotations_in_CDI_Stereotype.adoc [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+
+=== Dev Contacts
+
+* mailto:mstefank@redhat.com[Martin Stefanko]
+* mailto:asoldano@redhat.com[Alessio Soldano]
+* mailto:rsigal@redhat.com[Ron Sigal]
+
+=== QE Contacts
+
+* TBD
```
Comment (marekkopecky) [inline jaxrs/WFLY-8136_EAP7-1228_JAX-RS_annotations_in_CDI_Stereotype.adoc]: It's me :)

File: jaxrs/WFLY-8136_EAP7-1228_JAX-RS_annotations_in_CDI_Stereotype.adoc [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+
+=== Affected Projects or Components
+
+* RESTEasy
+
+=== Other Interested Projects
+
+== Requirements
+
+* RESTEasy will be able to process JAX-RS annotations in CDI Stereotype when run in
```
Comment (marekkopecky) [inline jaxrs/WFLY-8136_EAP7-1228_JAX-RS_annotations_in_CDI_Stereotype.adoc]: Is CDI environment always present in WF? Or is it possible that CDI environment won't be available in WF deployment?

**Comment by marekkopecky** [comment]
@xstefank Should this feature works on sub-deployments? What about stereotype annotations on proxies/MP-REST client interfaces?

File: jaxrs/WFLY-8136_EAP7-1228_JAX-RS_annotations_in_CDI_Stereotype.adoc (hunk 1 above)
Comment (marekkopecky) [reply jaxrs/WFLY-8136_EAP7-1228_JAX-RS_annotations_in_CDI_Stereotype.adoc]: Support of javax.ws.rs.NameBinding should be also clarified here ...

Comment (xstefank) [reply jaxrs/WFLY-8136_EAP7-1228_JAX-RS_annotations_in_CDI_Stereotype.adoc]: NameBinding is targeting only annotation_type. Stereotypes may be used only on classes and producer methods and fields.

File: jaxrs/WFLY-8136_EAP7-1228_JAX-RS_annotations_in_CDI_Stereotype.adoc (hunk 3 above)
Comment (xstefank) [reply jaxrs/WFLY-8136_EAP7-1228_JAX-RS_annotations_in_CDI_Stereotype.adoc]: @marekkopecky it wildfy it should be always present (at least up to my knowledge). However RESTEAsy can be used in environments/containers where CDI is not available.

File: jaxrs/WFLY-8136_EAP7-1228_JAX-RS_annotations_in_CDI_Stereotype.adoc (hunk 1 above)
Comment (marekkopecky) [reply jaxrs/WFLY-8136_EAP7-1228_JAX-RS_annotations_in_CDI_Stereotype.adoc]: Yes, user can create new annotation with NameBinding and user can try to use this new annotation on Stereotype ...

**Comment by xstefank** [comment]
@marekkopecky AD updated

File: jaxrs/WFLY-8136_EAP7-1228_JAX-RS_annotations_in_CDI_Stereotype.adoc [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+the CDI environment
+* RESTEasy will ignore JAX-RS annotations in CDI Stereotypes when no CDI environment 
+will be available
+
+=== Non-Requirements
+
+* As Stereotypes can be defined only at class, method and field levels, JAX-RS 
+annotations that cannot be applied at these places are not covered by this proposal. 
+This for instance covers https://docs.oracle.com/javaee/7/api/javax/ws/rs/NameBinding.html[@NameBinding]
+as it can be applied only to annotations (ElementType.ANNOTATION_TYPE).
```
Comment (marekkopecky) [inline jaxrs/WFLY-8136_EAP7-1228_JAX-RS_annotations_in_CDI_Stereotype.adoc]: What about this? 
```java
@NameBinding
@Retention(RUNTIME)
@Target({ METHOD, TYPE })
public @interface MyFilterBinding {
}
```

```java
@Provider
@MyFilterBinding
public class GreetingFilter implements ContainerRequestFilter {
    @Override
    public void filter(ContainerRequestContext requestContext) throws IOException {
        System.out.println("Filter");
    }
}
```

```java
@Stereotype
@Retention(RUNTIME)
@Target(TYPE)
@MyFilterBinding
public @interface CustomStereotype {
}
```

```java
@Path("/with")
@RequestScoped
@CustomStereotype
public class ResourceWith {
    @GET
    public String withInterceptor() {
        return "withInterceptor";
    }
}
```

**Comment by jamezp** [comment]
Should we close this? I don't really see this as an RFE since the spec requires the following:

[Section 11.2.3](https://jakarta.ee/specifications/restful-ws/3.1/jakarta-restful-ws-spec-3.1#cdi)
> In a product that supports CDI, implementations MUST support the use of CDI-style Beans as root resource classes, providers and Application subclasses. Providers and Application subclasses MUST be singletons or use application scope.

I've filed https://github.com/wildfly/wildfly/pull/17626 this. The Jakarta REST API itself should be adding the `@Stereotype` and proper scope to each required annotation type in its next release.

**Comment by xstefank** [comment]
@jamezp I think this can be closed. I leave it up to you. If it's landing in Jakarta REST no point in this RFE.

**Pull request closed without merging**